### ISSUE TITLE
Destroying emulator

### ISSUE THREAD

**Issue by filips123** [copy/v86#283]
Is it possible to destroy emulator programmatically? With destroying I mean removing all adapters (mouse and keyboard should be accessible again) and cleaning up memory.

This is useful if the site needs to load multiple systems, but not at the same time. Currently, the only possible solution is to reload the site.

**Comment by iamahuman** [comment]
`v86.reset();`

**Comment by filips123** [comment]
@iamahuman What does this do? It is not in docs.

I want to remove all adapters and cleanup memory.

**Comment by copy** [comment]
There's a partial implementation here: https://github.com/copy/v86/blob/eb736ab406c87e55e156beb7f889372a4dbfaf92/src/browser/starter.js#L459, but it needs to be extended to also destroy the other adapters. Pull requests are welcome.

**Comment by filips123** [comment]
I can submit PR for other adapters.
What adapters are used? Do all of them have `destroy` method?

Main problem with destroying is probably to clean up memory. How can this be done?

**Comment by copy** [comment]
> I can submit PR for other adapters.
> What adapters are used? Do all of them have destroy method?

Have a look at the code. IIRC most of them have a `destroy` but I'm not sure if they're complete. You'll need to check manually.

> Main problem with destroying is probably to clean up memory. How can this be done?

It should be garbage collected automatically as soon as all emulator objects are unreachable. You can check this in Chromiums developer tools by taking a heap snapshot in the memory tab and looking for an arraybuffer object that is larger than 128 MB.

**Comment by iamahuman** [comment]
@filips123 Are there any other references to ArrayBuffer objects? Even they are normal objects and any competent Javascript engine would garbage-collect it once it decides it can never be reached to by user code.

**Comment by filips123** [comment]
@iamahuman I think no. But maybe this is because adapters are still used.

**Comment by iamahuman** [comment]
@filips123 wait, I think I missed DOM event listeners whose closures do need to still refer to them. IMHO they can still be cut out by removing either the listeners with `removeEventListener` or the attached nodes altogether.

**Comment by iamahuman** [comment]
@filips123 I think the best way to figure it out would be to use a profiler that can enumerate all references to a specified object or value on runtime.

**Comment by filips123** [comment]
@iamahuman I currently don't have much time to do this. I will probably wait until #290 is merged to see if references are deleted.

---

If you also want to test this, I created simple emulator on [GitHib Gist](https://gist.github.com/filips123/1d0807f77b35182550c968ac40e66105).

And I also attached my memory snapshot files from Firefox and Chrome. I tested this with:

* Attaching some floppy image (Windows 1.01).
* Creating emulator.
* Starting emulation.
* Waiting a bit.
* Stopping emulation.
* Destroying emulator.
* Waiting a bit.
* Taking memory snapshot.

[Download memory snapshots](https://github.com/copy/v86/files/3233238/v86.Memory.Snapshots.zip)
